File: nchito-android/app/src/main/java/com/owenphiri/nchito/data/PricingService.kt
package com.owenphiri.nchito.data

/**
 * Fair-price bands — see nchito-ios/INNOVATION.md §5.1.
 *
 * Posting into a market with no reference prices leads to lowballing and slow,
 * hesitant posting. Quoting what comparable work actually settled at makes the
 * marketplace feel fair to both sides.
 *
 * Mirrors `price_band()` in 0002_phase1_defensibility.sql, and the iOS
 * PricingService, using the same linear-interpolation percentile so all three
 * agree on the same numbers.
 */
data class PriceBand(
    val low: Double,      // 25th percentile
    val median: Double,
    val high: Double,     // 75th percentile
    val sampleSize: Int,
    /** False when the local market was too thin and we widened to nationwide. */
    val isLocal: Boolean,
) {
    enum class Verdict(val message: String) {
        LOW("Below the going rate — you may get few applicants"),
        FAIR("In line with similar gigs"),
        HIGH("Above the going rate — expect fast applications"),
    }

    fun verdictFor(price: Double): Verdict = when {
        price < low -> Verdict.LOW
        price > high -> Verdict.HIGH
        else -> Verdict.FAIR
    }

    val rangeText: String get() = "${low.kwacha()}–${high.kwacha()}"

    val sourceText: String
        get() = "$sampleSize similar ${if (isLocal) "local " else ""}gig${if (sampleSize == 1) "" else "s"}"
}

object PricingService {
    /**
     * Below this, a sample says more about who happened to post than about the
     * market, so we widen the search rather than advise from noise.
     */
    const val MINIMUM_SAMPLE = 5

    /** Returns null when there isn't enough history — callers must show nothing. */
    fun band(category: GigCategory, city: String, gigs: List<Gig>): PriceBand? {
        val comparable = gigs.filter { it.category == category }
        val local = comparable.filter { it.city == city }
        val useLocal = local.size >= MINIMUM_SAMPLE
        val sample = if (useLocal) local else comparable

        if (sample.size < MINIMUM_SAMPLE) return null

        val prices = sample.map { it.payZMW }.sorted()
        return PriceBand(
            low = percentile(prices, 0.25),
            median = percentile(prices, 0.50),
            high = percentile(prices, 0.75),
            sampleSize = prices.size,
            isLocal = useLocal,
        )
    }

    /** Same method as Postgres's `percentile_cont`, so app and database agree. */
    private fun percentile(sorted: List<Double>, p: Double): Double {
        if (sorted.isEmpty()) return 0.0
        if (sorted.size == 1) return sorted[0]

        val position = p * (sorted.size - 1)
        val lowerIndex = position.toInt()
        val upperIndex = minOf(lowerIndex + 1, sorted.size - 1)
        val weight = position - lowerIndex

        val interpolated = sorted[lowerIndex] * (1 - weight) + sorted[upperIndex] * weight
        return Math.round(interpolated * 100) / 100.0
    }
}

File: nchito-ios/Nchito/Services/PricingService.swift
import Foundation

/// Fair-price bands — see INNOVATION.md §5.1.
///
/// Posting a gig in a market with no reference prices leads to lowballing and
/// slow, hesitant posting. Quoting what comparable work actually settled at
/// makes the marketplace feel fair to both sides and speeds up posting.
///
/// Mirrors `price_band()` in 0002_phase1_defensibility.sql. Production reads
/// the RPC; this computes the same quartiles locally over mock data.
struct PriceBand: Equatable {
    var low: Double       // 25th percentile
    var median: Double
    var high: Double      // 75th percentile
    var sampleSize: Int
    /// False when the local market was too thin and we widened to nationwide.
    var isLocal: Bool

    /// How a proposed price sits against the band.
    enum Verdict: Equatable {
        case low, fair, high

        var message: String {
            switch self {
            case .low:  return "Below the going rate — you may get few applicants"
            case .fair: return "In line with similar gigs"
            case .high: return "Above the going rate — expect fast applications"
            }
        }

        var icon: String {
            switch self {
            case .low:  return "arrow.down.circle.fill"
            case .fair: return "checkmark.circle.fill"
            case .high: return "arrow.up.circle.fill"
            }
        }
    }

    func verdict(for price: Double) -> Verdict {
        if price < low { return .low }
        if price > high { return .high }
        return .fair
    }

    var rangeText: String { "\(low.kwacha)–\(high.kwacha)" }

    var sourceText: String {
        "\(sampleSize) similar \(isLocal ? "local " : "")gig\(sampleSize == 1 ? "" : "s")"
    }
}

enum PricingService {
    /// Below this, a sample says more about who happened to post than about
    /// the market, so we widen the search rather than advise from noise.
    static let minimumSample = 5

    /// Returns nil when there simply isn't enough history — callers must show
    /// nothing rather than invent a number.
    static func band(for category: GigCategory, city: String, in gigs: [Gig]) -> PriceBand? {
        let comparable = gigs.filter { $0.category == category }

        let local = comparable.filter { $0.city == city }
        let useLocal = local.count >= minimumSample
        let sample = useLocal ? local : comparable

        guard sample.count >= minimumSample else { return nil }

        let prices = sample.map(\.payZMW).sorted()
        return PriceBand(
            low: percentile(prices, 0.25),
            median: percentile(prices, 0.50),
            high: percentile(prices, 0.75),
            sampleSize: prices.count,
            isLocal: useLocal
        )
    }

    /// Linear interpolation between ranks — the same method as Postgres's
    /// `percentile_cont`, so the app and the database agree.
    private static func percentile(_ sorted: [Double], _ p: Double) -> Double {
        guard !sorted.isEmpty else { return 0 }
        guard sorted.count > 1 else { return sorted[0] }

        let position = p * Double(sorted.count - 1)
        let lowerIndex = Int(position.rounded(.down))
        let upperIndex = Swift.min(lowerIndex + 1, sorted.count - 1)
        let weight = position - Double(lowerIndex)

        let interpolated = sorted[lowerIndex] * (1 - weight) + sorted[upperIndex] * weight
        return (interpolated * 100).rounded() / 100
    }
}
